google-cloud-spanner/src/main/java/com/google/cloud/spanner/connection/PostgreSQLStatementParser.java

```diff
@@ -149,30 +152,76 @@ ParametersInfo convertPositionalParametersToNamedParametersInternal(char paramCh
     return new ParametersInfo(paramIndex - 1, named.toString());
   }
 
-  private int skip(String sql, int currentIndex, StringBuilder result) {
+  /**
+   * Note: This is an internal API and breaking changes can be made without prior notice.
+   *
+   * <p>Returns the PostgreSQL-style query parameters ($1, $2, ...) in the given SQL string. The
+   * SQL-string is assumed to not contain any comments. Use {@link #removeCommentsAndTrim(String)}
+   * to remove all comments before calling this method. Occurrences of query-parameter like strings
+   * inside quoted identifiers or string literals are ignored.
+   *
+   * <p>The following example will return a set containing ("$1", "$2"). <code>
+   * select col1, col2, "col$4"
+   * from some_table
+   * where col1=$1 and col2=$2
+   * and not col3=$1 and col4='$3'
+   * </code>
+   *
+   * @param sql the SQL-string to check for parameters. Must not contain comments.
+   * @return A set containing all the parameters in the SQL-string.
+   */
+  @InternalApi
+  public Set<String> getQueryParameters(String sql) {
+    Preconditions.checkNotNull(sql);
+    int maxCount = countOccurrencesOf('$', sql);
+    Set<String> parameters = new HashSet<>(maxCount);
+    int currentIndex = 0;
+    while (currentIndex < sql.length() - 1) {
+      char c = sql.charAt(currentIndex);
+      if (c == '$' && Character.isDigit(sql.charAt(currentIndex + 1))) {
+        // Look ahead for the first non-digit. That is the end of the query parameter.
+        int endIndex = currentIndex + 2;
+        while (endIndex < sql.length() && Character.isDigit(sql.charAt(endIndex))) {
+          endIndex++;
+        }
+        parameters.add(sql.substring(currentIndex, endIndex));
+        currentIndex = endIndex;
+      } else {
+        currentIndex = skip(sql, currentIndex, null);
+      }
+    }
+    return parameters;
+  }
+
+  private int skip(String sql, int currentIndex, @Nullable StringBuilder result) {
     char currentChar = sql.charAt(currentIndex);
     if (currentChar == SINGLE_QUOTE || currentChar == DOUBLE_QUOTE) {
-      result.append(currentChar);
+      appendIfNotNull(result, currentChar);
```

Comment:
nit: will it ever be `null`?

Reply:
Yes, `result` will be null if the caller does not need the result. That's a (minor) performance optimization not to build a resulting query string when the call just wants to scan through the SQL string to extract specific information. So in this specific case we skip building a result string as we are only interested in extracting the query parameters, and not building a new SQL string with for example a different style query parameters (e.g. when converting JDBC style `?` parameters to PostgreSQL style parameters `$1`).